src/cache_cf.cc

```diff
@@ -298,139 +299,144 @@ skip_ws(const char* s)
-FindStatement(const char* line, const char* statement)
+/// extracts the next token if it matches the given keyword
+/// \returns whether the token was extracted
+static bool
+SkipKeyword(const SBuf &keyword, Parser::Tokenizer &tk)
 {
-    assert(line != nullptr);
-    assert(statement != nullptr);
-
-    const char* str = skip_ws(line);
-    unsigned len = strlen(statement);
-    if (strncmp(str, statement, len) == 0) {
-        str += len;
-        if (*str == '\0')
-            return str;
-        else if (xisspace(*str))
-            return skip_ws(str);
+    const auto savedTk = tk;
+    if (const auto token = ExtractOptionalToken(tk)) {
+        if (token == keyword)
+            return true;
     }
-
-    return nullptr;
+    tk = savedTk;
+    return false;
 }
 
-static bool
-StrToInt(const char* str, long& number)
+/// extracts the next input token (skipping any leading space) and interprets it as
+/// a signed integer (in decimal, hex, or octal base per Parser::Tokenizer::int64())
+static int64_t
+ExtractNumber(const char * const description, Parser::Tokenizer &tk)
 {
-    assert(str != nullptr);
-
-    char* end;
-    number = strtol(str, &end, 0);
-
-    return (end != str) && (*end == '\0'); // returns true if string contains nothing except number
+    auto numberParser = Parser::Tokenizer(ExtractToken(description, tk));
+    int64_t result = 0;
+    if (!numberParser.int64(result, 0, true))
+        throw TextException(ToSBuf("cannot parse ", description), Here());
+    if (!numberParser.remaining().isEmpty())
+        throw TextException(ToSBuf("found trailing garbage after parsing ", description), Here());
+    return result;
 }
 
 static bool
-EvalBoolExpr(const char* expr)
+EvalBoolExpr(Parser::Tokenizer &tk)
 {
-    assert(expr != nullptr);
-    if (strcmp(expr, "true") == 0) {
+    static const auto keywordTrue = SBuf("true");
+    if (tk.remaining() == keywordTrue)
         return true;
-    } else if (strcmp(expr, "false") == 0) {
+
+    static const auto keywordFalse = SBuf("false");
+    if (tk.remaining() == keywordFalse)
         return false;
-    } else if (const char* equation = strchr(expr, '=')) {
-        const char* rvalue = skip_ws(equation + 1);
-        char* lvalue = (char*)xmalloc(equation - expr + 1);
-        xstrncpy(lvalue, expr, equation - expr + 1);
-        trim_trailing_ws(lvalue);
 
-        long number1;
-        if (!StrToInt(lvalue, number1))
-            fatalf("String is not a integer number: '%s'\n", lvalue);
-        long number2;
-        if (!StrToInt(rvalue, number2))
-            fatalf("String is not a integer number: '%s'\n", rvalue);
+    const auto lhs = ExtractNumber("left-hand operand of an equality condition", tk);
 
-        xfree(lvalue);
-        return number1 == number2;
-    }
-    fatalf("Unable to evaluate expression '%s'\n", expr);
-    return false; // this place cannot be reached
+    const auto op = ExtractToken("equality sign in an equality condition", tk);
```

Comment:
New parsing code became a bit touchy - e.g., valid old configurations with no spaces `if ${process_number}=1` or two spaces `if ${process_number}  =  1` are not allowed. Are we going to adopt more rigid parsing rules here?

Reply:
> Are we going to adopt more rigid parsing rules here?

No, we should preserve the original grammar (to the extent it is practical). Thank you for flagging this change/bug! I hope I fixed it (as of branch commit 6d897ce0). I also added a few test cases (but they can be improved further by adding SMP macros).

Reply:
> I also added a few test cases (but they can be improved further by adding SMP macros).

I now added an SMP macro test and temporary workarounds for Tokenizer::int64() bugs (so that the tests pass while we wait for PR 1842).